File: sorting/LC45_jumping_game_2/solution.cpp
#include <iostream>
#include <algorithm>

class Solution {
public:
    int jump(std::vector<int>& nums) {
        int n = nums.size();
        // if the length of array is 1, no need to jump at all.
        if (n <= 1) {
            return 0;
        }

        int maxPos = nums[0];
        int maxSteps = nums[0];

        int jumps = 1;
        for (int i = 1; i < n; ++i) {
            if (maxSteps < i) {
                jumps++;
                maxSteps = maxPos;
            }
            maxPos = std::max(maxPos, nums[i] + i);
        }

        return jumps;
    }
};

int main() {
    Solution s;
    std::vector<int> nums = {2,3,1,1,4};
    std::cout << s.jump(nums) << std::endl;
}

File: sorting/LC45_jumping_game_2/solution.go
package main

import (
	"fmt"
)

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func jump(nums []int) int {
	n := len(nums)
	if n < 2 {
		return 0
	}

	maxPos := nums[0]
	maxSteps := nums[0]

	jumps := 1
	for i := 1; i < n; i++ {
		if maxSteps < i {
			jumps++
			maxSteps = maxPos
		}
		maxPos = max(maxPos, nums[i]+i)
	}

	return jumps
}

func main() {
	nums := []int{2, 3, 1, 1, 4}
	fmt.Println(jump(nums))
}
